GPflow/data_holders.py

```diff
@@ -0,0 +1,105 @@
+import tensorflow as tf
+import numpy as np
+
+from .param import DataHolder
+
+class TensorData(DataHolder):
+    def __init__(self, dataTensor):
+        DataHolder.__init__(self)
+        self._tf_array = dataTensor
+        
+    def get_feed_dict(self): 
+        #All done using a TensorFlow tensor so 
+        #there is no feed_dict associated with TensorData
+        return {}
+
+class DictData(DataHolder):
+    def __init__(self, array, on_shape_change='raise'):
+        """
+        array is a numpy array of data.
+        on_shape_change is one of ('raise', 'pass', 'recompile'), and
+        determines the behaviour when the data is set to a new value with a
+        different shape
+        """
+        DataHolder.__init__(self)
+        self._array = array
+        self._tf_array = tf.placeholder(dtype=self._array.dtype,
+                                        shape=[None]*self._array.ndim,
+                                        name=self.name)
+        assert on_shape_change in ['raise', 'pass', 'recompile']
+        self.on_shape_change = on_shape_change
+
+    def set_data(self, array):
+        """
+        Setting a data into self._array before any TensorFlow execution.
+        If the shape of the data changes, then either:
+         - raise an exception
+         - raise the recompilation flag.
+         - do nothing
+        according to the option in self.on_shape_change.
+        """
+        if self.shape == array.shape:
+            self._array[...] = array  # just accept the new values
+        else:
+            if self.on_shape_change == 'raise':
+                raise ValueError("The shape of this data must not change. \
+                                  (perhaps make the model again from scratch?)")
+            elif self.on_shape_change == 'recompile':
+                self._array = array.copy()
+                self.highest_parent._needs_recompile = True
+                if hasattr(self.highest_parent, '_kill_autoflow'):
+                    self.highest_parent._kill_autoflow()
```

Comment:
@alexggmatthews , this function gets called when the user assigns a new array to the object. If the new array is the same shape as the old array, nothing happens, we just accept the new data. If the shape has changed, that might be bad news for the model, which may have parameters whose shape matches that of the data (VGP, GPMC). In this case, we can either 1) raise an exception,  2) recompile the model or 3) do nothing

1) isnt used as far as I remember
2) is used for VGP and GPMC, which have special compile functions to cope
3) is used for SGPMC, SVGP, GPR, SGPR, which can all cope with the shape of the data being changed on-the-fly. 

edit: perhaps you were referring to the kill_autoflow switch? This removes any autoflow objects from the model, perhaps a little conservative. It's slightly messy, we should probably do that every time the _needs recompile flaga is set, automatically at the Parameterizsed or model level.